libsplinter/src/service/scabbard/state.rs

```diff
@@ -408,49 +414,57 @@ impl Events {
     ) -> Result<Self, ScabbardStateError> {
         let mut iter = Events {
             transaction_receipt_store,
-            start_id,
+            query: EventQuery::Fetch(start_id),
             cache: VecDeque::default(),
         };
         iter.reload_cache()?;
         Ok(iter)
     }
 
     fn reload_cache(&mut self) -> Result<(), ScabbardStateError> {
-        let transaction_receipt_store = self.transaction_receipt_store.read().map_err(|err| {
-            ScabbardStateError(format!("transaction receipt store lock poisoned: {}", err))
-        })?;
+        match self.query {
+            EventQuery::Fetch(ref start_id) => {
+                let transaction_receipt_store =
+                    self.transaction_receipt_store.read().map_err(|err| {
+                        ScabbardStateError(format!(
+                            "transaction receipt store lock poisoned: {}",
+                            err
+                        ))
+                    })?;
 
-        self.cache = if let Some(id) = &self.start_id {
-            transaction_receipt_store.iter_since_id(id.clone())
-        } else {
-            transaction_receipt_store.iter()
-        }
-        .map_err(|err| {
-            ScabbardStateError(format!(
-                "failed to get transaction receipts from store: {}",
-                err
-            ))
-        })?
-        .take(ITER_CACHE_SIZE)
-        .map(|ref receipt| receipt_into_scabbard_state_change_event(receipt))
-        .collect::<VecDeque<_>>();
-
-        self.start_id = Some(
-            self.cache
-                .back()
-                .map(|event| event.id.clone())
-                .unwrap_or_else(|| "".into()),
-        );
+                self.cache = if let Some(id) = start_id.as_ref() {
+                    transaction_receipt_store.iter_since_id(id.clone())
+                } else {
+                    transaction_receipt_store.iter()
+                }
+                .map_err(|err| {
+                    ScabbardStateError(format!(
+                        "failed to get transaction receipts from store: {}",
+                        err
+                    ))
+                })?
+                .take(ITER_CACHE_SIZE)
+                .map(|ref receipt| receipt_into_scabbard_state_change_event(receipt))
+                .collect::<VecDeque<_>>();
+
+                self.query = self
+                    .cache
+                    .back()
+                    .map(|event| EventQuery::Fetch(Some(event.id.clone())))
+                    .unwrap_or(EventQuery::Exhausted);
 
-        Ok(())
+                Ok(())
+            }
+            EventQuery::Exhausted => Ok(()),
+        }
     }
 }
 
 impl Iterator for Events {
     type Item = StateChangeEvent;
 
     fn next(&mut self) -> Option<Self::Item> {
-        if self.cache.is_empty() {
+        if self.cache.is_empty() && self.query != EventQuery::Exhausted {
```

Comment:
Can you add a unit test to verify this change?

Reply:
Added.